nixos/modules/config/stub-ld.nix

```diff
@@ -0,0 +1,81 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) last splitString getExe mkOption types mdDoc mkIf optionals;
+  inherit (builtins) elem toFile;
+
+  stub-ld = pkgs: let
+    message = ''
+      You seem to be trying to run a dynamically linked executable that
+      wasn't made for NixOS. This is possible, but not without extra steps,
+      because NixOS, by design, does not have a global library path.
+
+      There are several solutions available:
+
+      1. If the program is packaged in nixpkgs, use that version.
+
+      2. Write a nix expression for the program to package it
+         yourself. There are 3 approaches to this:
+
+         a. Build from source. This is ideal.
+
+         b. Fix the program's ELF metadata to hardcode paths to libraries
+            using autoPatchelfHook. Use this if building from source isn't
+            feasible.
+
+         c. Wrap the program to run in a normal-seeming container using
+            buildFHSEnv. This is undesirable, but sometimes necessary, for
+            example, if the package downloads and runs further executables.
+
+      3. If you expect to be doing this sort of thing a lot, you can set up
+         nix-ld using programs.nix-ld.enable. Make sure to add any needed
+         dependencies to programs.nix-ld.libraries.
+
+      4. Try running your program with steam-run, which uses the same
+         buildFHSEnv container as the steam package. This is quick and easy,
+         and usually works, but isn't a particularly good long-term solution.
+    '';
+
+    inherit (pkgs.stdenv) system;
+
+    libDir = if elem system [ "x86_64-linux" "mips64-linux" "powerpc64le-linux" ]
+             then "lib64"
+             else "lib";
+
+    ldsoBasename = last (splitString "/" pkgs.stdenv.cc.bintools.dynamicLinker);
+  in pkgs.pkgsStatic.runCommandCC "nixos-stub-ld" {} ''
+    ${getExe pkgs.unixtools.xxd} -i -n message <${toFile "message" message} >main.c
+    cat <<EOF >>main.c
+
+    #include <stdio.h>
+
+    int main() {
+      fwrite(message, sizeof(unsigned char), message_len, stderr);
+      return 1;
+    }
+    EOF
```

Comment:
Perhaps it'd be simpler to read the message from a file.

Reply:
Eh, not much simpler, honestly. And I like that it's self-contained.

Reply:
I think embedding into the binary is fine.